presto-client/src/main/java/com/facebook/presto/client/FixJsonDataUtils.java

```diff
@@ -96,36 +98,59 @@ private static Object fixValue(TypeSignature signature, Object value)
             return fixValue(signature.getDistinctTypeInfo().getBaseType(), value);
         }
         if (signature.getBase().equals(ARRAY)) {
-            List<Object> fixedValue = new ArrayList<>();
-            for (Object object : List.class.cast(value)) {
-                fixedValue.add(fixValue(signature.getTypeParametersAsTypeSignatures().get(0), object));
+            if (List.class.isAssignableFrom(value.getClass())) {
+                List<Object> fixedValue = new ArrayList<>();
+                for (Object object : List.class.cast(value)) {
+                    fixedValue.add(fixValue(signature.getTypeParametersAsTypeSignatures().get(0), object));
+                }
+                return fixedValue;
+            }
+            // JSON keys may be serialized and deserialized as simple strings
+            else if (value.getClass() == String.class) {
+                List<String> newValue = listJsonCodec(String.class).fromJson(value.toString());
+                return fixValue(signature, newValue);
             }
-            return fixedValue;
+            throw new IllegalArgumentException(String.format("Unexpected type found: %s", value.getClass()));
         }
         if (signature.getBase().equals(MAP)) {
-            TypeSignature keySignature = signature.getTypeParametersAsTypeSignatures().get(0);
-            TypeSignature valueSignature = signature.getTypeParametersAsTypeSignatures().get(1);
-            Map<Object, Object> fixedValue = new HashMap<>();
-            for (Map.Entry<?, ?> entry : (Set<Map.Entry<?, ?>>) Map.class.cast(value).entrySet()) {
-                fixedValue.put(fixValue(keySignature, entry.getKey()), fixValue(valueSignature, entry.getValue()));
+            if (Map.class.isAssignableFrom(value.getClass())) {
+                TypeSignature keySignature = signature.getTypeParametersAsTypeSignatures().get(0);
+                TypeSignature valueSignature = signature.getTypeParametersAsTypeSignatures().get(1);
+                Map<Object, Object> fixedValue = new HashMap<>();
+                for (Map.Entry<?, ?> entry : (Set<Map.Entry<?, ?>>) Map.class.cast(value).entrySet()) {
+                    fixedValue.put(fixValue(keySignature, entry.getKey()), fixValue(valueSignature, entry.getValue()));
+                }
+                return fixedValue;
             }
-            return fixedValue;
+            // JSON keys may be serialized and deserialized as simple strings
+            else if (value.getClass() == String.class) {
+                Map<String, String> newValue = mapJsonCodec(String.class, String.class).fromJson(value.toString());
```

Comment:
Please cache the codec in a static private final variable.

Reply:
Added codecs to private static variable